test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java

```diff
@@ -529,13 +532,28 @@ public void testToQuery() throws IOException {
         }
     }
 
+    // TODO: Deprecate this method?
     protected QueryBuilder rewriteQuery(QB queryBuilder, QueryRewriteContext rewriteContext) throws IOException {
         QueryBuilder rewritten = rewriteAndFetch(queryBuilder, rewriteContext);
         // extra safety to fail fast - serialize the rewritten version to ensure it's serializable.
         assertSerialization(rewritten);
         return rewritten;
     }
 
+    protected QueryBuilder rewriteQuery(
+        QB queryBuilder,
+        QueryRewriteContext coordinatorRewriteContext,
+        SearchExecutionContext shardRewriteContext
+    ) throws IOException {
+        // The first rewriteAndFetch call simulates rewriting on the coordinator node
+        // The second rewriteAndFetch call simulates rewriting on the shard
+        QueryBuilder rewritten = rewriteAndFetch(queryBuilder, coordinatorRewriteContext);
+        rewritten = rewriteAndFetch(rewritten, shardRewriteContext);
+        // extra safety to fail fast - serialize the rewritten version to ensure it's serializable.
+        assertSerialization(rewritten);
+        return rewritten;
+    }
```

Comment:
I needed to add a second version of `rewriteQuery` that simulates how queries are rewritten twice: once on the coordinator node and another time on the data shard.